lib/medtronic/processData.js

```diff
@@ -0,0 +1,378 @@
+    var carbInput = struct.extractByte(record.body,0);
+    var bgRaw = [struct.extractByte(record.head,1)];
+    bgRaw = bgRaw.concat(struct.extractByte(record.body,1) & 0x0f);
+    var bgInput = struct.extractShort(bgRaw,0);
+    var carbRatio = struct.extractByte(record.body,14) / 10.0;
+    var bgTarget = {};
+    bgTarget.low = struct.extractByte(record.body,5);
+    var isf = struct.extractByte(record.body,4);
+    bgTarget.high = struct.extractByte(record.body,3);
+    var recommended = {};
+    recommended.carb = struct.extractBEShort(record.body,7) / 40.0;
+    recommended.correction = struct.extractByte(record.body,6) / 40.0;
+    recommended.net = struct.extractBEShort(record.body,12) / 40.0;
+    var iob =  struct.extractBEShort(record.body,10) / 40.0;
+
+    var wizard = cfg.builder.makeWizard()
+      .with_recommended({
+        carb: recommended.carb,
+        correction: recommended.correction,
+        net: recommended.net
+      })
+      .with_bgInput(bgInput)
+      .with_carbInput(carbInput)
+      .with_insulinOnBoard(iob)
+      .with_insulinCarbRatio(carbRatio)
+      .with_insulinSensitivity(isf)
+      .with_bgTarget({
+        low: bgTarget.low,
+        high: bgTarget.high
+      })
+      .with_units('mg/dL'); //TODO: see if we can find units
+
+      wizard = wizard.with_deviceTime(sundial.formatDeviceTime(record.jsDate))
+          .set('index', record.index)
+          .set('jsDate', record.jsDate);
+
+      cfg.tzoUtil.fillInUTCInfo(wizard, record.jsDate);
+      postrecords.push(wizard);
+
+  });
+
+  return postrecords;
+};
+
+var buildBolusRecords = function (records) {
+  var bolusRecords = filterHistory([RECORD_TYPES.BOLUS], records);
+  var postrecords = [];
+
+  bolusRecords.forEach(function(bolus) {
+    console.log('Bolus:',bolus.jsDate, common.bytes2hex(bolus.head), common.bytes2hex(bolus.date));
+  });
+
+  for(var i=0; i < bolusRecords.length; i++) {
+    var record = bolusRecords[i];
+
+    var bolus;
+
+    if (larger) {
+      var duration = record.head[7] * 30 * sundial.MIN_TO_MSEC;
+      var amount = struct.extractBEShort(record.head,3)/40.0;
+      var programmed = struct.extractBEShort(record.head,1)/40.0;
+      var iob = struct.extractBEShort(record.head,5)/40.0;
+      if (duration > 0) {
+        //TODO: put iob in payload?
+
+        if((record.date[2] & 0xC0) == 0x80) {
+          bolus = cfg.builder.makeDualBolus()
+            .with_duration(duration)
+            .with_extended(amount);
+
+            if(programmed !== amount) {
+              // dual bolus was cancelled
+              var actualDuration = Math.round((amount / (programmed * 1.0)) * duration);
+              bolus = bolus.with_expectedExtended(programmed)
+                        .with_expectedDuration(duration)
+                        .with_duration(actualDuration);
+            }
+
+            i+=1; // advance to next bolus for normal portion
```

Comment:
So as mentioned before, I think this sequencing (or it may be the sequencing in relation to wizard events, not sure if affects both or just one) can be interrupted. How do we know/check here that after advancing we really do have a bolus record, or the right bolus record? Perhaps the situation is actually way better than I thought! Do you get the `bolusRecords` all in one batch, so they are all guaranteed to be boluses? The problem we had in CareLink parsing was that we were dealing with all types at once and sometimes an alarm would intervene in the sequence (by log index).